File: Solutions/0775/0775.c
/***************************************************************************************************
 * Copyright: washing
 * FileName: 0775.c
 * Author: washing
 * Version: 1.0
 * Date: 2022-11-16 10:39:29
***************************************************************************************************/

bool isIdealPermutation(int* nums, int numsSize){
    if (numsSize == 1) return true;
    int mi = 0;
    if (nums[0] == 0 || nums[1] == 0) {
        mi++;
        if (nums[0] == 1 || nums[1] == 1) mi++;
    }
    for (int idx=0; idx < numsSize; idx++) {
        if (nums[idx] > mi) return false;
        while (1) {
            if (nums[idx] == mi) mi++;
            else if (idx+1 < numsSize) {
                if (nums[idx+1] == mi) mi++;
                else if (idx+2 < numsSize && nums[idx+2] == mi) mi++;
                else break;
            }
            else break;
        }
    }
    return true;
}

File: Solutions/0775/0775.py
# -*- coding:utf-8 -*-
# Author: washing
# DateTime: 2022/11/16 10:24
# File: 0775.py
# Desc: 

class Solution:
    def isIdealPermutation(self, nums: List[int]) -> bool:
        if len(nums) == 1: return True
        mi = 0
        if 0 in nums[:2]:
            mi += 1
            if 1 in nums[:2]:
                mi += 1
        for idx,i in enumerate(nums):
            # print(idx,i,mi)
            if i > mi: return False
            while mi in nums[idx:idx+3]: mi+=1
        return True
